query/physicalplan/physicalplan.go

```diff
@@ -233,25 +246,85 @@ func Build(ctx context.Context, pool memory.Allocator, tracer trace.Tracer, s *d
 			}
 			return false
 		case plan.Projection != nil:
-			p, err := Project(pool, tracer, plan.Projection.Exprs)
-			if err != nil {
-				return false
+			// For each previous physical plan create one Projection
+			for i := 0; i < len(prev); i++ {
+				p, err := Project(pool, tracer, plan.Projection.Exprs)
+				if err != nil {
+					return false
+				}
+				phyPlans = append(phyPlans, p)
 			}
-			phyPlans = append(phyPlans, p)
 		case plan.Distinct != nil:
-			phyPlans = append(phyPlans, Distinct(pool, tracer, plan.Distinct.Exprs))
+			concurrency := concurrencyHardcoded
+			if len(prev) > 1 {
+				// Just in case the previous plan's concurrency differs from the hardcoded concurrency
+				concurrency = len(prev)
+			}
+
+			phyPlans = make([]PhysicalPlan, 0, concurrency)
+			for i := 0; i < concurrency; i++ {
+				phyPlans = append(phyPlans, Distinct(pool, tracer, plan.Distinct.Exprs))
+			}
+
+			// If the prev is 1 then we need a synchronizer to handle the concurrency
+			// writing into a single callback of the previous PhysicalPlan.
+			if len(prev) == 1 {
```

Comment:
If we decide to make concurrency configurable at runtime (which we might want to do based on load), I think this will create an unnecessary synchronizer if we have concurrency == 1 for any reason. What do you think of creating a struct with more planning state to make more informed decisions of what to plan? For example, store some sort of notion of whether we *have* to plan a synchronizer. Maybe it's fine to also check `len(phyPlans)`, but it seems to me like this state complexity will start to balloon. This struct I suggest might also be useful for more planning information (e.g. ordering column guarantees)